src/hex.rs

```diff
@@ -186,7 +186,6 @@ impl ToHex for [u8] {
     }
 }
 
-#[cfg(any(test, feature = "std"))]
```

Comment:
Wouldn't it be easier to just have a pub function that converts these objects to hex and then only have the `FromHex`/etc traits on std-mode, that way its usable either way, but we don't take a new dependency just for hex?

Reply:
If you are referring to the new dependency `core2`, that is for Read/Write traits used in impls.rs

If you are referring to alloc, yes it may be worthy to avoid using alloc just for this traits

Reply:
I'm referring to `core2` (indeed, alloc isn't really a "new" dependency, as its part of std anyway) - an alternative approach to avoid `core2` would be to reimplement the `io::Write/Read` traits as functions, with versions which expose `Write/Read` if we're building with `std`. Ultimately a user not using std isn't going to have `Write` or `Read` anyway, they just want a byte buffer (or a string, in the case of hex), which we can just give them, instead of exposing the `Write` or `Read` traits (which they're unlikely to use anyway - they don't have std so probably no files, just writing into Vecs).

Reply:
#122 is not using `core2` and it is simply using `input` function from the Engine

Reply:
@TheBlueMatt I don't think it's realistic to expect people not to use `Read` or `Write`, these are universal traits used by all sorts of things, and it looks like `core2` is the "standard" way to do this without std. So I think we have to bite the bullet and just use it, even though it bumps our MSRV for nostd and even though it's an extra dep.

We could plausibly get away without it here, or maybe feature-gate it even, but for rust-bitcoin I think we'll need it no matter what, unless we want to either disable all the decoding/encoding stuff (i.e. most of the library) or rewrite it.

Reply:
Honestly, I'm not sure why we'd really care? Even with std, we wanted an additional function so just built a new trait for Rust-Lightning. It's 15 LoC with documentation to just define a trait and implement it for std (and, presumably, for core2::Write with a feature), but avoids an entire dependency for those who just access the underlying [u8; 32] (which I think Rust-Lightning does even for serialization). https://github.com/rust-bitcoin/rust-lightning/blob/main/lightning/src/util/ser.rs#L44-L58

Reply:
Ah, yeah, I see what you're saying. If some 3rd party writes a `fn myfunc<W: Write>(thing: W)` and we have elsewhere `impl<T: OurCustomNoStdWrite> Write for T { ... }` then people can use our types when calling `myfunc` and not care that we did our own weird thing.

@RCasatta I've changed my mind to agree with Matt now.

Reply:
rust-bitcoin would need to replace every `std::io::Write` with our `OurCustomNoStdWrite` from bitcoin-hashes, is it correct?

(rust-bech32 doesn't use Write, otherwise, it would be somehow out of context here)

Reply:
Yeah, this is correct. Similar to what we did had with `Encodable` prior to replacing that with `io::Write`

Reply:
Interesting discussion.  I didn't see this before opening the following PR in rust-bitcoin with a similar approach:

https://github.com/rust-bitcoin/rust-bitcoin/pull/603/files#diff-76866598ce8fd16261a27ac58a84b2825e6e77fc37c163a6afa60f0f4477e569R28

Reply:
(note that I rename it to just `Write` when I import it into the root - in case you are looking at the rest of the PR)